internal/lru/lru.go

```diff
@@ -34,64 +38,120 @@ type Cache struct {
 	// executed when an entry is purged from the cache.
 	OnEvicted func(key string, value interface{})
 
-	ll    *list.List
-	cache map[string]*list.Element
+	window time.Duration
+	ll     *list.List
+	cache  map[string]*list.Element
+	mu     sync.RWMutex
 }
 
 type entry struct {
-	key   string
-	value interface{}
+	key       string
+	value     interface{}
+	timestamp time.Time
 }
 
 // New creates a new Cache.
 // If maxEntries is zero, the cache has no limit and it's assumed
 // that eviction is done by the caller.
-func New(maxEntries int) *Cache {
+// If window is zero the "windowed updates" of the LRU list will be disabled.
+// For further information on this technique see: https://www.openmymind.net/High-Concurrency-LRU-Caching/
+func New(maxEntries int, window time.Duration) *Cache {
 	return &Cache{
 		MaxEntries: maxEntries,
 		ll:         list.New(),
 		cache:      make(map[string]*list.Element),
+		window:     window,
 	}
 }
 
 // Add adds a value to the cache.
-func (c *Cache) Add(key string, value interface{}) {
-	if c.cache == nil {
-		c.cache = make(map[string]*list.Element)
-		c.ll = list.New()
-	}
-	if ee, ok := c.cache[key]; ok {
-		c.ll.MoveToFront(ee)
-		ee.Value.(*entry).value = value
+func (c *Cache) Add(key string, val interface{}) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if ele, ok := c.cache[key]; ok {
+		c.ll.MoveToFront(ele)
+		ele.Value.(*entry).value = val
 		return
 	}
-	ele := c.ll.PushFront(&entry{key, value})
+	c.addLocked(key, val)
+}
+
+func (c *Cache) addLocked(key string, val interface{}) {
+	timestamp := time.Now()
+	ent := &entry{key, val, timestamp}
+	ele := c.ll.PushFront(ent)
 	c.cache[key] = ele
 	if c.MaxEntries != 0 && c.ll.Len() > c.MaxEntries {
-		c.RemoveOldest()
+		c.removeOldestLocked()
 	}
 }
 
 // Get looks up a key's value from the cache.
 func (c *Cache) Get(key string) (value interface{}, ok bool) {
-	if c.cache == nil {
-		return
```

Comment:
You could extract the write-locked version of the lookup to a function and special case `c.window == 0` to call that directly.